pkg/daemon/cleanup/cleanup.go

```diff
@@ -0,0 +1,173 @@
+func removeCMServices(c client.Client, ctx context.Context, svcMgr manager.Manager, cm *core.ConfigMap) error {
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	// Build up log message and failures
+	servicesRemoved := []string{}
+	var failedRemovals error
+	// The services are ordered by increasing priority already, so stop them in reverse order to avoid dependency issues
+	for i := len(cmData.Services) - 1; i >= 0; i-- {
+		service := cmData.Services[i]
+		if err := svcMgr.DeleteService(service.Name); err != nil {
+			failedRemovals = errors.Errorf("unable to delete %s (%v); %s", service.Name, err, failedRemovals)
+		}
+		servicesRemoved = append(servicesRemoved, service.Name)
+	}
+	if failedRemovals != nil {
+		return failedRemovals
+	}
+	klog.Infof("deleted services %v from ConfigMap %s", servicesRemoved, cm.Name)
+	return nil
+}
+
+// outOfSyncCleanup removes all services on the instance with the Openshift managed tag, as well as any services defined
+// in the latest retreivable version of the services ConfigMap
+func outOfSyncCleanup(c client.Client, ctx context.Context, svcMgr manager.Manager) error {
+	klog.Infof("cleaning up all %s Windows services", windows.ManagedTag)
+	allServices, err := svcMgr.GetServices()
+	if err != nil {
+		return errors.Wrap(err, "could not determine existing Windows services")
+	}
+	// Build up any failures
+	var failedRemovals error
+	for service := range allServices {
+		// Re-query existing services to avoid processing one that has already been removed
+		currServices, err := svcMgr.GetServices()
+		if err != nil {
+			return err
+		}
+		if _, ok := currServices[service]; !ok {
+			continue
+		}
+
+		winSvcObj, err := svcMgr.OpenService(service)
+		if err != nil {
+			failedRemovals = errors.Errorf("unable to open %s (%v); %s", service, err, failedRemovals)
+			continue
+		}
+		defer winSvcObj.Close()
+
+		config, err := winSvcObj.Config()
+		if err != nil {
+			failedRemovals = errors.Errorf("unable to get %s service config (%v); %s", service, err, failedRemovals)
+			continue
+		}
+		// Only remove the service if it has the expected tag
+		if !strings.HasPrefix(config.Description, windows.ManagedTag) {
+			continue
+		}
+		if err = svcMgr.DeleteServiceAndDependents(service); err != nil {
+			failedRemovals = errors.Errorf("unable to delete %s (%v); %s", service, err, failedRemovals)
+		}
+	}
+	if failedRemovals != nil {
+		// Since this is a best effort cleanup, warn but do not return error on failure
+		klog.Warning(failedRemovals)
+	}
+
+	// Try to get the latest services ConfigMap. If it exists, the services in it will be deconfigured.
+	// This will enable clean up instances even if service descriptions were somehow overwritten.
+	cm, err := getLatestServicesCM(c, ctx)
+	if err != nil {
+		klog.Warningf("unable to find latest services ConfigMap: %s", err.Error())
+		return nil
+	}
+	return removeCMServices(c, ctx, svcMgr, cm)
```

Comment:
Reversing the order of this cleanup and the rest of the function makes more sense to me.
This is giving us a list of services that we know we installed. We should be handling that before combing through the rest of the services.